BedrockServer.cpp

```diff
@@ -2256,45 +2259,47 @@ void BedrockServer::handleSocket(Socket&& socket, bool fromControlPort, bool fro
+                        atomic<bool> finished = false;
 
-                        function<void()> callback = [&m, &cv]() {
+                        function<void()> callback = [&m, &cv, &finished]() {
                             // Lock the mutex above (which will be locked by this thread while we're queuing), which waits
                             // for `handleSocket` to release it's lock (by calling `wait`), and then notify the waiting
                             // socket thread.
                             lock_guard lock(m);
+                            finished = true;
                             cv.notify_all();
                         };
 
-                        // Ok, none of above synchronization code gets called unless the command has a socket to respond
-                        // on.
+                        // Ok, none of above synchronization code gets called unless the command has a socket to respond on.
                         bool hasSocket = command->socket;
                         if (hasSocket) {
                             // Set the destructor callback for when the command finishes.
                             command->destructionCallback = &callback;
-
-                            // And lock the mutex so the command can't complete until we are in `wait` below.
-                            lock.lock();
                         }
 
-                        // Now we'll queue this command in one of two queues.
+                        // Now we queue or run this command.
                         auto _syncNodeCopy = atomic_load(&_syncNode);
                         if (_syncNodeCopy && _syncNodeCopy->getState() == SQLiteNode::STANDINGDOWN) {
                             _standDownQueue.push(move(command));
                         } else {
-                            SINFO("Queuing new '" << command->request.methodLine << "' command from local client, with "
-                                  << _commandQueue.size() << " commands already queued.");
-                            _commandQueue.push(move(command));
+                            // If this command is scheduled in the future, we can't just run it, we need to enqueue it to run at that point.
+                            // This functionality will go away as we remove the queues from bedrock, and so this can be removed at that time.
+                            if (command->request.calcU64("commandExecuteTime") > STimeNow()) {
+                                // These are implicitly fire-and-forget commands and a response will already have been sent in `buildCommandFromRequest`.
+                                _commandQueue.push(move(command));
+                                continue;
+                            } else {
+                                SINFO("Running new '" << command->request.methodLine << "' command from local client, with " << _commandQueue.size() << " commands already queued.");
```

Comment:
Fun fact: renaming this log from `Queuing new` to `Running new` probably broke the "Auth Requests" graph in grafana, as we had a regex looking for that here: https://github.com/Expensify/Salt/blob/373a4551d7ef81489f91acc7346580d20a023047/shared_pkgs/bedrock/files/logsToStatsd.pl#L172-L189